File: web/circleci_verify.py
import hmac


def verify_signature(secret, headers, body):
    # get the v1 signature from the `circleci-signature` header
    signature_from_header = {
        k: v
        for k, v in [
            pair.split("=") for pair in headers["circleci-signature"].split(",")
        ]
    }["v1"]

    # Run HMAC-SHA256 on the request body using the configured signing secret
    valid_signature = hmac.new(
        bytes(secret, "utf-8"), bytes(body, "utf-8"), "sha256"
    ).hexdigest()

    # use constant time string comparison to prevent timing attacks
    return hmac.compare_digest(valid_signature, signature_from_header)

File: web-rs/src/circleci_verify.rs
use hex;
use hmac::{Hmac, Mac};
use sha2::Sha256;
use warp::{reject, Filter};

const SIGNATURE_VERSION: &'static str = "v1";
const HEADER: &'static str = "circleci-signature";

pub fn extract_compatible_signature(header: &str) -> Option<&str> {
    header
        .split(",")
        .find_map(|pair| match pair.split("=").collect::<Vec<&str>>()[..] {
            [v, sig] if v == SIGNATURE_VERSION => Some(sig),
            _ => None,
        })
}

pub fn verify_signature(secret: &str, signature: &str, body: &[u8]) -> bool {
    println!(
        "secret: {}, signature: {}, body: {:?}",
        secret, signature, body
    );

    let mut h = Hmac::<Sha256>::new_from_slice(secret.as_bytes()).unwrap();
    h.update(body);
    h.verify(hex::decode(signature).unwrap().as_slice().into())
        .is_ok()
}

#[derive(Debug)]
struct InvalidSignature;
impl reject::Reject for InvalidSignature {}

#[derive(Debug)]
struct IncompatibleSignatureVersion;
impl reject::Reject for IncompatibleSignatureVersion {}

pub fn verify_filter(secret: &str) -> impl Filter<Extract = (), Error = warp::Rejection> + Clone + '_ {
    warp::header::<String>(HEADER)
        .and(warp::body::bytes())
        .and_then(move |header: String, body: warp::hyper::body::Bytes| {
            let signature = extract_compatible_signature(&header);
            match signature {
                None => futures::future::err(reject::custom(IncompatibleSignatureVersion)),
                Some(signature) => {
                    if verify_signature(secret, signature, &body) {
                        futures::future::ok(())
                    } else {
                        futures::future::err(reject::custom(InvalidSignature))
                    }
                }
            }
        })
        .untuple_one()
}
